src/core_plugins/kibana/public/management/sections/indices/edit_index_pattern/source_filters_table/source_filters_table.js

```diff
@@ -1,128 +1,205 @@
-                const matcher = fieldWildcardMatcher([ filter.value ]);
-                // compute which fields match a filter
-                const matches = $scope.indexPattern.getNonScriptedFields().map(f => f.name).filter(matcher).sort();
-                if ($scope.$parent.fieldFilter && !angularFilter(matches, $scope.$parent.fieldFilter).length) {
-                  return;
-                }
-                // compute the rows
-                const rowScope = $scope.$new();
-                rowScope.filter = filter;
-                rowScopes.push(rowScope);
-                $scope.rows.push([
-                  {
-                    markup: filterHtml,
-                    scope: rowScope
-                  },
-                  size(matches) ? escape(matches.join(', ')) : '<em>The source filter doesn\'t match any known fields.</em>',
-                  {
-                    markup: controlsHtml,
-                    scope: rowScope
-                  }
-                ]);
-              });
-              // Update the tab count
-              find($scope.$parent.editSections, { index: 'sourceFilters' }).count = $scope.rows.length;
-            }
-          });
-        }
+  constructor(props) {
+    super(props);
 
-        all() {
-          return this.$scope.indexPattern.sourceFilters || [];
-        }
+    // Source filters do not have any unique ids, only the value is stored.
+    // To ensure we can create a consistent and expected UX when managing
+    // source filters, we are assigning a unique id to each filter on the
+    // client side only
+    this.clientSideId = 0;
 
-        delete(filter) {
-          const doDelete = () => {
-            if (this.editing === filter) {
-              this.editing = null;
-            }
-
-            this.$scope.indexPattern.sourceFilters = without(this.all(), filter);
-            return this.save();
-          };
-
-          const confirmModalOptions = {
-            confirmButtonText: 'Delete',
-            onConfirm: doDelete,
-            title: 'Delete source filter?'
-          };
-          confirmModal('', confirmModalOptions);
-        }
+    this.state = {
+      filterToDelete: undefined,
+      isDeleteConfirmationModalVisible: false,
+      isSaving: false,
+      filters: [],
+      ...DEFAULT_TABLE_OF_RECORDS_STATE,
+    };
+  }
 
-        create() {
-          const value = this.newValue;
-          this.newValue = null;
-          this.$scope.indexPattern.sourceFilters = [...this.all(), { value }];
-          return this.save();
-        }
+  componentWillMount() {
+    this.updateFilters();
+  }
 
-        save() {
-          this.saving = true;
-          this.$scope.indexPattern.save()
-            .then(() => this.editing = null)
-            .catch(notify.error)
-            .finally(() => this.saving = false);
-        }
+  updateFilters = () => {
+    const filters = this.props.indexPattern.sourceFilters.map(filter => ({
```

Comment:
I think we need to add some logic to handle the case when `sourceFilters` is undefined. I'm getting this error:

![image](https://user-images.githubusercontent.com/1238659/36510913-90a76e30-1719-11e8-8641-dac86f321a5e.png)

Reply:
This seems to work:

```js
    const filters = this.props.indexPattern.sourceFilters ? this.props.indexPattern.sourceFilters.map(filter => ({
      ...filter,
      clientId: ++this.clientSideId,
    })) : [];
```